dbt/include/synapse/macros/materializations/snapshot/strategies.sql

```diff
@@ -0,0 +1,7 @@
+{% macro synapse__snapshot_hash_arguments(args) %}
+    {{ return(sqlserver__snapshot_hash_arguments(args)) }}
+{% endmacro %}
+
+{% macro snapshot_check_strategy(node, snapshotted_rel, current_rel, config, target_exists) %}
+    {{ return(snapshot_check_strategy(node, snapshotted_rel, current_rel, config, target_exists)) }}
+{% endmacro %}
```

Comment:
looks like [we're overwriting this macro directly in `dbt-sqlserver`](https://github.com/dbt-msft/dbt-sqlserver/blob/throwaway/dbt/include/sqlserver/macros/materializations/snapshot/strategies.sql)

1. does that mean that this chunk is still needed?
2. should this be a dispatched macro elsewhere?

Reply:
If you need to directly reimplement this core macro [`snapshot_check_strategy`](https://github.com/fishtown-analytics/dbt/blob/30f473a2b1e261e3f4b712564df854f8a95134f3/core/dbt/include/global_project/macros/materializations/snapshot/strategies.sql#L133) for SQLServer/Synapse, then I think it should be a dispatched macro. I don't think you'd be able to overwrite it as-is within a plugin (either `dbt-sqlserver` or `dbt-synapse`) since they share the `dbt` namespace for macros

Reply:
looking at the diffs (left:`core` right:`sql-server1) there seem to be three main differences:
1. some updates in core that haven't yet propogated down to `dbt-sqlserver` (sidenote: just the thought of keeping up with changes like this long-term exhausts me)
    1. https://github.com/fishtown-analytics/dbt/pull/2001
    2. https://github.com/fishtown-analytics/dbt/pull/2749
    3. https://github.com/fishtown-analytics/dbt/pull/1994
2. replacing `snapshot_hash_arguments()` with `sqlserver__snapshot_hash_arguments()` (must have been a hack just before `adapter.dispatch` came through
3. [one line](https://github.com/dbt-msft/dbt-sqlserver/blob/master/dbt/include/sqlserver/macros/materializations/snapshot/strategies.sql#L25) where `and not` is used instead of `!=`
    `({{ snapshotted_rel }}.{{ col }} is null) and not ({{ current_rel }}.{{ col }} is null)`


if the 3rd diff isn't necessary (I fear it will be) then we can arguably get away with `core`'s versions. if it isn't, perhaps I make a PR to `core` that dispatches the single line where things break to ameliorate the `sqlserver`'s code footprint (and help us rest more easily)?


<img width="1769" alt="image" src="https://user-images.githubusercontent.com/8158673/103996918-54afb680-514f-11eb-9711-9ace56ab21e7.png">

Reply:
update -- yep it's gotta be `and not`. which upon reflection, kind of scares me because this `sql-server` is saying `A & ~B` instead of `core`'s logic of `A` and `B` can't be both `True` or both `False`. @mikaelene any recollection as to why this is different?
https://github.com/dbt-msft/dbt-sqlserver/pull/79

Reply:
Hi the change from `!=` to `and not` is required. Don't really understand why but that is the reason we can't use the standard macro. Is this something we could change in core without wrecking the other adapters?

Reply:
> If you need to directly reimplement this core macro [`snapshot_check_strategy`](https://github.com/fishtown-analytics/dbt/blob/30f473a2b1e261e3f4b712564df854f8a95134f3/core/dbt/include/global_project/macros/materializations/snapshot/strategies.sql#L133) for SQLServer/Synapse, then I think it should be a dispatched macro. I don't think you'd be able to overwrite it as-is within a plugin (either `dbt-sqlserver` or `dbt-synapse`) since they share the `dbt` namespace for macros

I think this is because the snapshot macros were not really updated to the new dispatched thing when I implemented this. So this was the way to implement adapter specific macro for this. Maybe that has changed?

Reply:
> Maybe that has changed?
nope -- still the case! i'd still prefer that we macro-ize just that one line, because we missed so many good PRs in the last year alone!

Reply:
It wold make the core macro harder to read. Can we rewrite it to work on SQL Server and all the other databases without changing the logic?

Reply:
I agree that's the ideal outcome

Reply:
> if it isn't, perhaps I make a PR to `core` that dispatches the single line where things break to ameliorate the sqlserver's code footprint (and help us rest more easily)?

Ok, I think this is the right move, then.

I totally defer to both of you on this, but I don't quite follow how `and not` is working here to produce the same logioc.

If both `A` and `B` are null:
- `dbt-core` logic says: `True != True` which returns `False`
- `dbt-sqlserver` logic says: `True and not True` which returns `False`

If both `A` and `B` are not null:
- `dbt-core` logic says: `False != False` which returns `False`
- `dbt-sqlserver` logic says: `False and not False` which returns `False`

If `A` is null and `B` is not null:
- `dbt-core` logic says: `True != False` which returns `True`
- `dbt-sqlserver` logic says: `True and not False` which returns `True`

**HOWEVER If `A` is not null and `B` is null:**
- `dbt-core` logic says: `False != True` which returns `True`
- `dbt-sqlserver` logic says: `False and not True` which returns `False`... am I missing something on this one?

Reply:
Maybe like this? 
```select

  scenario,
  (a is null) != (b is null) as dbt_core_logic,
  (a is null) and not (b is null) or (b is null) and not (a is null) as dbt_sqlserver_logic

from (

  select 1 as scenario, null as a, null as b
  union all
  select 2 as scenario, 1 as a, 1 as b
  union all
  select 3 as scenario, null as a, 1 as b
  union all
  select 4 as scenario, 1 as a, null as b

) sbq
```

Reply:
i've made two PR to see if it works on:
- SQL Server https://github.com/dbt-msft/dbt-sqlserver/pull/79 (it does!), and
- dbt-core https://github.com/fishtown-analytics/dbt/pull/3000 (3000th PR!!!)

Reply:
update. it works!!!